doc/man/man3/DwmPcap.hh.3
.TH "DwmPcap.hh" 3 "libDwm-0.0.20240716" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DwmPcap.hh \- \fBDwm::Pcap\fP class definition\&.  

.SH SYNOPSIS
.br
.PP
\fR#include <pcap\&.h>\fP
.br
\fR#include <map>\fP
.br
\fR#include <string>\fP
.br

.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBDwm::Pcap\fP"
.br
.RI "This is a pcap wrapper/utility class\&. "
.in -1c
.SH "Detailed Description"
.PP 
\fBDwm::Pcap\fP class definition\&. 


.SH "Author"
.PP 
Generated automatically by Doxygen for libDwm-0\&.0\&.20240716 from the source code\&.


ChangeLog
2024-07-25  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as libDwm-0.9.45

	* README.md: mention libfmt dependency for the case where
	std::format is not available.

	* tagged as libDwm-0.9.44

	* classes/include/DwmGZIOCapable.hh: Documentation updates.
	* classes/include/DwmBZ2IOCapable.hh: Documentation updates.
	* classes/include/DwmASIOCapable.hh: Documentation updates.
	* classes/include/DwmDescriptorIOCapable.hh: Documentation updates.
	* classes/include/DwmFileIOCapable.hh: Documentation updates.
	* classes/include/DwmStreamIOCapable.hh: Documentation updates.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmDocMainPage.hh, README.md,
	doc/examples/DwmIOExample3.cc: Documentation updates.

2024-07-24  dwm  <dwm@nomnom.mcplex.net>

	* tagged as libDwm-0.9.43

	* mkrelease.sh: 'push --tags' should be 'push --follow-tags'

	* tagged as libDwm-0.9.42

	* classes/include/DwmSysLogger.hh: Fix pasto on platforms where
	we need libfmt (std:: should be fmt::), namely Raspberry Pi OS.

2024-07-24  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.41

	* classes/tests/TestThreadQueue.cc: Use max concurrency on
	multicore systems when running performance tests, just to
	add to the stress.  Also use C++14 digit separator (') in
	some numeric literals for readability.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/src/DwmSysLogger.cc, classes/include/DwmSysLogger.hh:
	Allow use of std::source_location, and use them in the Syslog
	and FSyslog macros.

	* Removed the following files since I don't need them here:

	classes/include/DwmHostPinger.hh
	classes/include/DwmIpv4IcmpDestinationUnreachable.hh
	classes/include/DwmIpv4IcmpEchoReply.hh
	classes/include/DwmIpv4IcmpEchoRequest.hh
	classes/include/DwmIpv4IcmpMessage.hh
	classes/include/DwmIpv4IcmpRedirect.hh
	classes/include/DwmIpv4IcmpSourceQuench.hh
	classes/include/DwmIpv4IcmpTimeExceeded.hh
	classes/include/DwmIpv4PacketHeader.hh
	classes/include/DwmIpv4TcpHeader.hh
	classes/include/DwmIpv4TcpPacket.hh
	classes/include/DwmIpv4TcpPayload.hh
	classes/include/DwmIpv4UdpHeader.hh
	classes/include/DwmIpv4UdpPacket.hh
	classes/include/DwmIpv4UdpPayload.hh
	classes/include/DwmPatricia.hh
	classes/include/DwmPingDestination.hh
	classes/include/DwmPingDestinationEntry.hh
	classes/include/DwmSocket.hh
	classes/include/DwmXmlAttribute.hh
	classes/include/DwmXmlElement.hh
	classes/src/DwmHostPinger.cc
	classes/src/DwmIpv4IcmpDestinationUnreachable.cc
	classes/src/DwmIpv4IcmpEchoReply.cc
	classes/src/DwmIpv4IcmpEchoRequest.cc
	classes/src/DwmIpv4IcmpMessage.cc
	classes/src/DwmIpv4IcmpRedirect.cc
	classes/src/DwmIpv4IcmpSourceQuench.cc
	classes/src/DwmIpv4IcmpTimeExceeded.cc
	classes/src/DwmIpv4PacketHeader.cc
	classes/src/DwmIpv4TcpHeader.cc
	classes/src/DwmIpv4TcpPacket.cc
	classes/src/DwmIpv4TcpPayload.cc
	classes/src/DwmIpv4UdpHeader.cc
	classes/src/DwmIpv4UdpPacket.cc
	classes/src/DwmIpv4UdpPayload.cc
	classes/src/DwmPingDestination.cc
	classes/src/DwmPingDestinationEntry.cc
	classes/src/DwmSocket.cc
	classes/src/DwmXmlAttribute.cc
	classes/src/DwmXmlElement.cc
	classes/src/Makefile
	classes/tests/Makefile
	classes/tests/TestDwmSocket.cc
	classes/tests/TestHostPinger.cc
	classes/tests/TestIpv4IcmpDestinationUnreachable.cc
	classes/tests/TestIpv4IcmpEchoReply.cc
	classes/tests/TestIpv4IcmpEchoRequest.cc
	classes/tests/TestIpv4IcmpRedirect.cc
	classes/tests/TestIpv4IcmpSourceQuench.cc
	classes/tests/TestIpv4IcmpTimeExceeded.cc
	classes/tests/TestIpv4PacketHeader.cc
	classes/tests/TestIpv4TcpHeader.cc
	classes/tests/TestIpv4TcpPacket.cc
	classes/tests/TestIpv4UdpHeader.cc
	classes/tests/TestIpv4UdpPacket.cc
	classes/tests/TestPatricia.cc
	classes/tests/ip.pcap
	classes/tests/tcp.pcap
	classes/tests/udp.pcap

2024-07-23  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* getvers.sh: bump shared library version to reflect ABI changes
	(vtables changed from removal of inheritance).

	* README.md: Document Read and Write requirements for
	ASIO with boost::asio::generic::stream_protocol::socket

	* classes/include/DwmIpv4IcmpMessage.hh,
	  classes/include/DwmIpv4PacketHeader.hh,
	  classes/include/DwmIpv4TcpHeader.hh,
	  classes/include/DwmIpv4TcpPacket.hh,
	  classes/include/DwmIpv4TcpPayload.hh,
	  classes/include/DwmIpv4UdpHeader.hh,
	  classes/include/DwmIpv4UdpPacket.hh,
	  classes/include/DwmIpv4UdpPayload.hh:
	No longer need to inherit from *Readable and *Writable classes.

	* classes/include/DwmASIOCapable.hh,
	  classes/include/DwmBZ2IOCapable.hh,
	  classes/include/DwmDescriptorIOCapable.hh,
	  classes/include/DwmFileIOCapable.hh,
	  classes/include/DwmGZIOCapable.hh,
	  classes/include/DwmStreamIOCapable.hh:
	Minor cleanup to concepts.

	* tagged as libDwm-0.9.40

	* doc/.gitignore: ignore libDwm.tag (generated by doxygen)

	* README.md: Add mkfbsdmnfst and mkdebcontrol as build
	dependencies.

2024-07-22  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Removed IN-PROGRESS file.

	* tagged as libDwm-0.9.39

	* Added README.md for the github mirror.

2024-07-21  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.38

2024-07-21  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* doc/footer.html: Tweak to point at my web site.

	* classes/src/DwmMacAddress.cc: Use sizeof() where appropriate.

	* Remove doxygen-generated documentation from repository; it's
	build output that doesn't need to be in the repository.

2024-07-20  Daniel McRobb  <dwm@toto.mcplex.net>

	* tagged as libDwm-0.9.37

2024-07-20  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* classes/src/DwmProcessTableLex.lex,
	classes/src/DwmProcessTableParse.y: 'ps' output on FreeBSD shows
	a single hyphen ('-') instead of a ctime() string for lstart for
	some processes.  Patch this by substituting the lstart from the
	nearest ancestor of the process that has a valid lstart.

	* classes/include/DwmBootTime.hh, classes/include/DwmUptime.hh
	classes/src/DwmBootTime.cc, classes/src/DwmUptime.cc:

classes/src/DwmPcap.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2005-2007, 2016, 2020
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file DwmPcap.cc
//!  \author Daniel W. McRobb
//!  \brief Pcap class implementation
//---------------------------------------------------------------------------

#include "DwmPortability.hh"

#include <cassert>
#include <cstdlib>
#include <cstring>

#include "DwmLocalInterfaces.hh"
#include "DwmSysLogger.hh"
#include "DwmPcap.hh"

extern "C" {
#if (defined(HAVE_NET_PFVAR_H) && defined(HAVE_NET_IF_PFLOG_H))
  #include <net/if.h>
  #include <net/pfvar.h>
  #include <net/if_pflog.h>
#endif
}

using namespace std;

namespace Dwm {

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  Pcap::Pcap()
      : _pcap(0), _errbuf()
  { }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  Pcap::~Pcap()
  {
    if (_pcap)
      Close();
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  int	Pcap::SetBufferSize(int bufferSize)
  {
    if (_pcap != nullptr) {
      return pcap_set_buffer_size(_pcap, bufferSize);
    }
    else {
      return PCAP_ERROR_ACTIVATED;
    }
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool Pcap::OpenLive(const string & device, int snaplen,
                      bool promisc, int timeout_msecs)
  {
    string  intfName = device;
    bool    rc = true;
    char    errbuf[PCAP_ERRBUF_SIZE];

    if (intfName.empty()) {
      map<string,LocalInterface>  interfaces;
      if ((! GetLocalInterfaces(interfaces))
          || interfaces.empty()) {
        Syslog(LOG_ERR, "Pcap::OpenLive() failed to find an interface");
        return(false);
      }
      map<string,LocalInterface>::const_iterator  i;
      for (i = interfaces.begin(); i != interfaces.end(); ++i) {
        if (i->second.IsUp() && i->second.IsRunning()) {
          intfName = i->first;
          break;
        }
      }
      if (intfName.empty()) {
        Syslog(LOG_ERR, "Pcap::OpenLive(): no suitable interface found");
        return(false);
      }
    }
      
    _pcap = pcap_open_live((char *)intfName.c_str(), snaplen, promisc,
                           timeout_msecs, errbuf);
    if (! _pcap) {
      rc = false;
      _errbuf = errbuf;
      Syslog(LOG_ERR, "Pcap::OpenLive(\"%s\") failed", intfName.c_str());
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool Pcap::SetDirection(pcap_direction_t direction)
  {
    bool  rc = false;
    if (_pcap) {
      if (pcap_setdirection(_pcap, direction) == 0) {
        rc = true;
      }
    }
    return rc;
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool Pcap::OpenOffline(const string & fname)
  {
    bool  rc = true;
    char  errbuf[PCAP_ERRBUF_SIZE];
    if (_pcap)
      Close();
    _pcap = pcap_open_offline(fname.c_str(), errbuf);
    if (!_pcap) {
      rc = false;
      _errbuf = errbuf;
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool Pcap::SetFilter(const string & filter, bool optimize)
  {
    bool  rc = false;

    if (! filter.empty()) {
      //  sigh, libpcap is not const-correct
      char *filt = (char *)strdup(filter.c_str());
      
      struct bpf_program  bpfProg;
      memset(&bpfProg, 0, sizeof(bpfProg));
      if (pcap_compile(_pcap, &bpfProg, filt, optimize, 0) >= 0) {
        if (pcap_setfilter(_pcap, &bpfProg) == 0) {
          rc = true;
        }
        pcap_freecode(&bpfProg);
      }
      free(filt);
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool Pcap::Close()
  {
    bool  rc = false;
    if (_pcap) {
      pcap_close(_pcap);
      _pcap = 0;
      rc = true;
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  int Pcap::Dispatch(int count, pcap_handler callback, uint8_t *user)
  {
    assert(_pcap);
    return(pcap_dispatch(_pcap, count, callback, user));
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  int Pcap::Loop(int count, pcap_handler callback, uint8_t *user)
  {
    return(pcap_loop(_pcap, count, callback, user));
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  void Pcap::BreakLoop()
  {
    return pcap_breakloop(_pcap);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  const string & Pcap::LastError() const
  {
    return(_errbuf);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  int Pcap::DataLinkOffset() const
  {
    int  rc = -1;
    std::map<int,int>::const_iterator  iter = 
      _dataLinkOffsets.find(pcap_datalink(_pcap));
    if (iter != _dataLinkOffsets.end())
      rc = iter->second;
    return(rc);
  }
  
  std::map<int,int>  Pcap::_dataLinkOffsets = {
#if (defined(DLT_PFLOG) && defined(PFLOG_HDRLEN))
    { DLT_PFLOG,   PFLOG_HDRLEN },
#endif
    { DLT_NULL,     4 },
    { DLT_EN10MB,  14 },
    { DLT_IEEE802, 22 },
    { DLT_FDDI,    21 },
    { DLT_LOOP,    4 }
  };
  
}  // namespace Dwm


//---------------------------- emacs settings -----------------------------
//  Local Variables:
//  mode: C++
//  tab-width: 2
//  indent-tabs-mode: nil
//  c-basic-offset: 2
//  End:
//-------------------------------------------------------------------------
